NCPCore/src/main/java/fr/neatmonster/nocheatplus/checks/moving/player/SurvivalFly.java

```diff
@@ -794,11 +698,177 @@ else if (resetFrom) {
+                limitFCMH = 1.34;
+            }
+            else if (data.liftOffEnvelope == LiftOffEnvelope.LIMIT_LIQUID 
+                    || data.liftOffEnvelope == LiftOffEnvelope.LIMIT_NEAR_GROUND) {
+                // limitFCMH = 1.05; // Seems to work on 1.10
+                limitFCMH = 1.1; // 1.8.8 in-water moves with jumping near/on surface. 1.2 is max factor for one move (!).
+                // TODO: Version+context dependent setting and/or confine by in-water moves, whatever.
+            }
+            else {
+                limitFCMH = 1.0;
+            }
+            // TODO: Configurable / adjust by medium type.
+            // TODO: Instead of velocityJumpPhase account for friction directly?
+            // TODO: Fly-NoFly + bunny-water transitions pose issues.
+            if (fcmh > limitFCMH && !data.isVelocityJumpPhase()) {
+                hDistanceAboveLimit = Math.max(hDistanceAboveLimit, (hDistance * (fcmh - limitFCMH)));
+                tags.add("hacc");
+                // Reset for now.
+                data.combinedMediumHCount = 0;
+                data.combinedMediumHValue = 0.0;
+            }
+            else {
+                // TODO: Other cases (1.0, between, ...)?
+                data.combinedMediumHCount = 1;
+                data.combinedMediumHValue = fcmhv;
+            }
+        }
+        return hDistanceAboveLimit;
+    }
+    
+  /**
+    * Directly checks for certain cheat implementation types that aim at walking on water
+    * 
+    * @param hDistance
+    * @param yDistance
+    * @param thisMove
+    * @param lastMove
+    * @return
+    */
+    private double waterWalkChecks(final MovingData data, final Player player, double hDistance, double yDistance, 
+                                               final PlayerMoveData thisMove, final PlayerMoveData lastMove,
+                                               final boolean fromOnGround, double hDistanceAboveLimit,
+                                               final boolean toOnGround, final PlayerLocation from, final PlayerLocation to){
+
+        if (hDistanceAboveLimit <= 0D && hDistance > 0.1D && yDistance == 0D && lastMove.toIsValid && lastMove.yDistance == 0D
+            && !toOnGround && !fromOnGround  
+            && BlockProperties.isLiquid(to.getTypeId()) 
+            && BlockProperties.isLiquid(from.getTypeId())
+            && !from.isHeadObstructed() && !to.isHeadObstructed() 
+            && !Bridge1_13.isSwimming(player) 
+            ) {
+            hDistanceAboveLimit = Math.max(hDistanceAboveLimit, hDistance);
+            tags.add("waterwalk");
+        }
+            
+        final Block blockUnder = player.getLocation().subtract(0, 0.3, 0).getBlock();
+        final Material blockAbove = player.getLocation().add(0, 0.10, 0).getBlock().getType();
+	    
+        if (blockUnder != null && blockAbove != null && blockUnder.getType().toString().endsWith("WATER") && blockAbove.name().endsWith("AIR")) {
```

Comment:
Would be nice to use BlockProperties#isAir and BlockProperties#isLiquid. It might be faster than string processing. Sorry for creating such mess, sometimes I'm quite silly.

Reply:
No worries lol. Glad to see you back :p